packages/flutter/test/material/stepper_test.dart

```diff
@@ -953,4 +954,100 @@ void main() {
 
      expect(material.elevation, 2.0);
    });
+
+  testWidgets('Stepper horizontal preserves state', (WidgetTester tester) async {
+    Widget buildFrame() {
+      return const MaterialApp(
+        home: Scaffold(
+          body: Center(
+            // Must break this out into its own widget purely to be able to call `setState()`
+            child: StatefulStepper(),
+          ),
+        ),
+      );
+    }
+
+    final Widget widget = buildFrame();
+    await tester.pumpWidget(widget);
+
+    // We are on step 1
+    expect(find.text('Step 2 Content'), findsNothing);
+
+    ColoredBox findBox() => tester.widget<ColoredBox>(
+      find.descendant(of: find.byKey(const Key('randomized-color')), matching: find.byType(ColoredBox)),
+    );
+    final Color color = findBox().color;
+
+    await tester.tap(find.text('Step 2'));
+    await tester.pumpAndSettle();
+
+    // Confirm that we flipped over to step 2
+    expect(find.text('Step 2 Content'), findsOneWidget);
+
+    await tester.tap(find.text('Step 1'));
+    await tester.pumpAndSettle();
+
+    // Confirm that we flipped back to step 1
+    expect(find.text('Step 2 Content'), findsNothing);
+    final Color color2 = findBox().color;
+    expect(color, equals(color2));
+  });
+}
+
+class StatefulStepper extends StatefulWidget {
+  const StatefulStepper({ Key? key }) : super(key: key);
+
+  @override
+  StatefulStepperState createState() => StatefulStepperState();
+}
+
+class StatefulStepperState extends State<StatefulStepper> {
+  int index = 0;
+
+  @override
+  Widget build(BuildContext context) {
+    return Stepper(
+      onStepTapped: (int i) => setState(() => index = i),
+      currentStep: index,
+      type: StepperType.horizontal,
+      steps: const <Step>[
+        Step(
+          title: Text('Step 1'),
+          content: RandomColorWidget(
+            key: Key('randomized-color'),
+          ),
+        ),
+        Step(
+          title: Text('Step 2'),
+          content: Text('Step 2 Content'),
+        ),
+      ],
+    );
+  }
+}
+
+class RandomColorWidget extends StatefulWidget {
```

Comment:
Rather than a random color pick, you could just bake in the states, like start step 1 with blue, call the state setter to make it red and then validate it is still red when you come back?